### DISCUSSION TITLE
Moving away from signal-driven I/O in Maiko

### DISCUSSION THREAD

**Discussion by waywardmonkeys** [Interlisp/medley#125] [General]
The usage of signal-driven I/O presents some issues:

* It isn't well supported on all platforms. In particular, Cygwin doesn't support it (perhaps well, perhaps not at all). Emscripten  doesn't support signals.
* It is part of Maiko having a CPU-saturating main loop and moving to polling for events would be an improvement.

I haven't really started to look at what's necessary here. Has anyone else?

I suspect that it will mean changes to the core interpreter loop in `src/xc.c`.

**Comment by masinter** [comment]
The main interpreter loop was THE hotspot in all performance testing. When we were first doing maiko for SPARC, their compiler was awful; did no optimizations. This led to the awful hack of a lex-on-asm peephole "optimizer", written in a way that it would be harmless when Sun fixed their compiler.  I'd be surprised if it actually did anything now.
Beyond that, the next hotspots were the small opcodes POP COPY IVARn PVARn to those would be a shame -- they  don't have any byteswap penalty. I imagined a macro for fast opcode ending which would fetch the next byte and dispatch to the next opcode with each "fast" opcodes.  Maiko doesn't actually USE signals except to write a bit somewhere that a signal occurred. (or it didn't -- this is all from memory)

**Reply by nbriggs** [reply]
> Maiko doesn't actually USE signals except to write a bit somewhere that a signal occurred. (or it didn't -- this is all from memory)

When it takes a `SIGIO` signal (handled by `getsignaldata()`)it grabs any incoming ethernet, keyboard or mouse events and processes them into the key state bit vectors, and handles any other I/O that is noticed via a `select()` on the open descriptors.  Other than that, there's some convoluted logic checking on interrupt requests at each JUMP opcode, which if I'm following the code correctly would be the only other time that I/O would get noticed.

**Reply by nbriggs** [reply]
You might remember that we found that when the code was incorrectly setting up the FD_SETs, the display would come up but the system was deaf to the keyboard.
